config/config.go

```diff
@@ -0,0 +1,213 @@
+	var errs []error
+	for key := range c {
+		if _, ok := params[key]; !ok {
+			errs = append(errs, fmt.Errorf("%q: %w", key, ErrUnrecognizedParameter))
+		}
+	}
+	return errs
+}
+
+// validateParamType validates that a parameter value is parsable to its assigned type.
+func (c Config) validateParamType(key string, param Parameter) error {
+	value := c[key]
+	// empty value is valid for all types
+	if c[key] == "" {
+		return nil
+	}
+
+	//nolint:exhaustive // type ParameterTypeFile and ParameterTypeString don't need type validations (both are strings or byte slices)
+	switch param.Type {
+	case ParameterTypeInt:
+		_, err := strconv.Atoi(value)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not an integer: %w", key, value, ErrInvalidParameterType)
+		}
+	case ParameterTypeFloat:
+		_, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not a float: %w", key, value, ErrInvalidParameterType)
+		}
+	case ParameterTypeDuration:
+		_, err := time.ParseDuration(value)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not a duration: %w", key, value, ErrInvalidParameterType)
+		}
+	case ParameterTypeBool:
+		_, err := strconv.ParseBool(value)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not a boolean: %w", key, value, ErrInvalidParameterType)
+		}
+	}
+	return nil
+}
+
+// validateParamValue validates that a configuration value matches all the
+// validations required for the parameter.
+func (c Config) validateParamValue(key string, param Parameter) error {
+	value := c[key]
+	var errs []error
+
+	isRequired := false
+	for _, v := range param.Validations {
+		if _, ok := v.(ValidationRequired); ok {
+			isRequired = true
+		}
+		err := v.Validate(value)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error validating %q: %w", key, err))
+			continue
+		}
+	}
+	if value == "" && !isRequired {
+		return nil // empty optional parameter is valid
+	}
+
+	return errors.Join(errs...)
+}
+
+// DecodeInto copies configuration values into the target object.
+// Under the hood, this function uses github.com/mitchellh/mapstructure, with
+// the "mapstructure" tag renamed to "json". To rename a key, use the "json"
+// tag. To embed structs, append ",squash" to your tag. For more details and
+// docs, see https://pkg.go.dev/github.com/mitchellh/mapstructure.
+func (c Config) DecodeInto(target any) error {
+	dConfig := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           &target,
+		DecodeHook: mapstructure.ComposeDecodeHookFunc(
+			emptyStringToZeroValueHookFunc(),
+			mapstructure.StringToTimeDurationHookFunc(),
+			mapstructure.StringToSliceHookFunc(","),
+		),
+		TagName: "json",
+		Squash:  true,
+	}
+
+	decoder, err := mapstructure.NewDecoder(dConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create decoder: %w", err)
+	}
+	err = decoder.Decode(c.breakUp())
```

Comment:
Why do we break up the config map? Is it because of `Squash: true` in the DecodeConfig above and embedded structs?

Reply:
I think it's because we want to be able to populate composed structs. You know, like when you have a global config struct and you embed it in your source config struct. This way mapstructure will populate the inner struct, otherwise it doesn't do that magic.